pkg/authenticator/ecrsecret.go

```diff
@@ -0,0 +1,134 @@
+package authenticator
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"fmt"
+	"os"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	varPackagesNamespace = "eksa-packages"
+	varConfigMapName     = "ns-secret-map"
+	varECRTokenName      = "ecr-token"
+)
+
+type ecrSecret struct {
+	clientset    kubernetes.Interface
+	nsReleaseMap map[string]string
+}
+
+var _ Authenticator = (*ecrSecret)(nil)
+
+func NewECRSecret(config *rest.Config) (*ecrSecret, error) {
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	cm, err := clientset.CoreV1().ConfigMaps(varPackagesNamespace).
+		Get(context.TODO(), varConfigMapName, metav1.GetOptions{})
```

Comment:
This constructor is doing a lot of things.

In general, the purpose of a constructor is to initialize any required memory fields within its struct. It's not really meant to be making network calls, or performing advanced logic. That sort of thing is best put in some sort of Init() method or the like.

The thing returned by a constructor should be safe from nil-pointer dereferences, not necessarily "ready to use", if that makes sense.

In this case, I suspect it might be better to let the caller of this initializer perform this ConfigMaps call, and pass the results into the constructor.

If it's too complicated or nuanced a thing for the caller to do, consider having another function, perhaps "SimpleSecretsAuthenticator" or something that could combine a few calls as necessary.

Reply:
done by just initializing it to empty and filing it later as it gets updated anytime the config map is changed